DBZ-7822 Add JDBC connection idle timeouts (#5522)
debezium/debezium: documentation/modules/ROOT/pages/connectors/jdbc.adoc

When this options is set, when an incoming event is for a topic called `orders`, and the destination database dialect is Oracle, the connector creates a table with the name `orders`, because the constructed SQL defines the name of the table as `"orders"`.
Enabling quoting results in the same behavior when the connector creates column names.

[[jdbc-connection-idle-timeouts]]
=== Connection Idle Timeouts

The JDBC sink connector for {prodname} leverages a connection pool to enhance performance.

> Connection pools are engineered to establish an initial set of connections, maintain a specified number of connections, and efficiently allocate them to the application as required.

Reviewer: ```suggestion
Connection pools are engineered to establish an initial set of connections, maintain a specified number of connections, and efficiently allocate connections to the application as required.
```
_"the application"_ being Debezium (and specifically the JDBC connector)?

Author: In this context, _application_ loosely means anything as I was describing connection pools and their purpose from a general purpose.

However, a challenge arises when connections linger idle in the pool, potentially triggering timeouts if they remain inactive beyond the configured idle timeout threshold of the database.

> To mitigate this issue, connection pools offer a mechanism to periodically validate the connection's activity.

Reviewer: ```suggestion
To mitigate the potential for idle connection threads to trigger timeouts, connection pools offer a mechanism that periodically validates the activity of each connection.
```

> This validation ensures that connections remain active and prevent them from being perceived as idle by the database.

Reviewer: ```suggestion
This validation ensures that connections remain active, and prevents the database from flagging them as idle.
```

> In case of a network disruption, attempting to utilize a terminated connection prompts the pool to replace it with a fresh one.

Reviewer: ```suggestion
In the event of a network disruption, if {prodname} attempts to use a terminated connection, the connector prompts the pool to replace it with a fresh one.
```
Is it Debezium/the connector the prompts for a new connection?

Author: Yes, _connector_ is right, _connection_ was incorrect.

> By default, the {prodname} JDBC sink connector does not conduct idle timeout tests. Nevertheless, this behavior can be easily configured through the settings:

Reviewer: ```suggestion
By default, the {prodname} JDBC sink connector does not conduct idle timeout tests. 
However, you can configure the connector to perform timeout tests by setting the following configuration property:
```

.Example timeout configuration
[source,json]
----
{
  "hibernate.c3p0.idle_test_period": "300"
}
----

The {prodname} JDBC sink connector uses the Hibernate C3P0 connection pool.

> By passing any C3P0 configuration using the `hibernate.c3p0.*` configuration namespace, the pool can be customized.

Reviewer: ```suggestion
You can customize the CP30 connection pool by setting properties in the hibernate.c3p0.*` configuration namespace.
```

> For instance, in this scenario, the `hibernate.c3p0.idle_test_period` configuration is used to define the duration between idle timeout tests, specified in seconds.

Reviewer: ```suggestion
For example, the preceding timeout configuration sets the `hibernate.c3p0.idle_test_period` property, to specify that the connector waits 300 seconds between idle timeout tests.
```

> Setting this value to 300 ensures that the connection pool assesses unused connections every five minutes.

Reviewer: ```suggestion
After you apply the configuration, the connection pool begins to assess unused connections every five minutes.
```

When this options is set, when an incoming event is for a topic called `orders`, and the destination database dialect is Oracle, the connector creates a table with the name `orders`, because the constructed SQL defines the name of the table as `"orders"`.
Enabling quoting results in the same behavior when the connector creates column names.

[[jdbc-connection-idle-timeouts]]
=== Connection Idle Timeouts

The JDBC sink connector for {prodname} leverages a connection pool to enhance performance.
Connection pools are engineered to establish an initial set of connections, maintain a specified number of connections, and efficiently allocate connections to the application as required.
However, a challenge arises when connections linger idle in the pool, potentially triggering timeouts if they remain inactive beyond the configured idle timeout threshold of the database.

To mitigate the potential for idle connection threads to trigger timeouts, connection pools offer a mechanism that periodically validates the activity of each connection.
This validation ensures that connections remain active, and prevents the database from flagging them as idle.

> In the event of a network disruption, if {prodname} attempts to use a terminated connection, the connector prompts the pool to replace it with a fresh one.

Reviewer: ```suggestion
In the event of a network disruption, if {prodname} attempts to use a terminated connection, the connector prompts the pool to generate a new connection.
```

By default, the {prodname} JDBC sink connector does not conduct idle timeout tests.

> However, you can configure the connector to perform timeout tests by setting the following configuration property:

Reviewer: As in my question below, is it the connector that performs the timeout tests, or is it that the connector can pass a configuration property to the connection pool to affect its behavior?
```suggestion
However, you can configure the connector to request the pool to perform timeout tests at a specified interval by setting the `hibernate.c3p0.idle_test_period` property.
For example:
```

.Example timeout configuration
[source,json]
----
{
  "hibernate.c3p0.idle_test_period": "300"
}
----

The {prodname} JDBC sink connector uses the Hibernate C3P0 connection pool.
You can customize the CP30 connection pool by setting properties in the hibernate.c3p0.*` configuration namespace.

> For example, the preceding timeout configuration sets the `hibernate.c3p0.idle_test_period` property, to specify that the connector waits 300 seconds between idle timeout tests.

Reviewer: ```suggestion
In the preceding example, the setting of the `hibernate.c3p0.idle_test_period` property causes the connector to wait 300 seconds between idle timeout tests.
```

> After you apply the configuration, the connection pool begins to assess unused connections every five minutes.

Reviewer: I'm questioning this sentence and the wording of L299. Does `hibernate.c3p0.idle_test_period` cause the connector to request the pool to conduct a timeout test at a specified interval? Or is this a pass-through property that directly configures the pool to set an interval for validating unused connections?  
Maybe we can replace L299-300 with the following text:

_In the preceding example, the setting of the `hibernate.c3p0.idle_test_period` property configures the connection pool to perform idle timeout tests every 300 seconds._

Author: I think this suggestion makes more sense.  It's simply a pass-through property that causes the pool to perform an idle test every 300` seconds in the example's use case, so what you've said here checks.
